drivers/at_cmd/at_cmd.c

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <logging/log.h>
+#include <zephyr.h>
+#include <stdio.h>
+#include <net/socket.h>
+#include <init.h>
+#include <bsd_limits.h>
+
+#include <at_cmd.h>
+
+LOG_MODULE_REGISTER(at_cmd, CONFIG_AT_CMD_LOG_LEVEL);
+
+#define RESPONSE_MAX_LEN  CONFIG_AT_CMD_RESPONSE_MAX_LEN
+#define THREAD_PRIORITY   K_PRIO_PREEMPT(CONFIG_AT_CMD_THREAD_PRIO)
+
+#define AT_CMD_OK_STR    "OK"
+#define AT_CMD_ERROR_STR "ERROR"
+#define AT_CMD_CMS_STR   "+CMS:"
+#define AT_CMD_CME_STR   "+CME:"
+
+static K_THREAD_STACK_DEFINE(socket_thread_stack, 1024);
+static K_MUTEX_DEFINE(cmd_pending);
+
+K_MSGQ_DEFINE(return_code_msq, sizeof(enum at_cmd_rc), 2, 4);
+
+static int              common_socket_fd;
+static int              last_error;
+static char             *response_buf;
+static u32_t            response_buf_len;
+
+static struct k_thread  socket_thread;
+static at_cmd_handler_t notification_handler;
+static at_cmd_handler_t current_cmd_handler;
+
+struct callback_work_item {
+	struct k_work    work;
+	char             data[RESPONSE_MAX_LEN];
+	at_cmd_handler_t callback;
+};
+
+static int open_socket(void)
+{
+	common_socket_fd = socket(AF_LTE, 0, NPROTO_AT);
+
+	if (common_socket_fd == -1) {
+		LOG_ERR("AT socket creation failed");
+		return -EFAULT;
+	}
+
+	return 0;
+}
+
+static enum at_cmd_rc get_return_code(char *buf)
+{
+	enum at_cmd_rc return_code = AT_CMD_NOTIFICATION;
+	char *tmpstr = NULL;
+
+	do {
+		tmpstr = strstr(buf, AT_CMD_OK_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_OK;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_ERROR_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_ERROR;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_CMS_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_CMS;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_CME_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_CME;
+			break;
+		}
+	} while (0);
+
+	if ((return_code == AT_CMD_CMS) || (return_code == AT_CMD_CME)) {
+		last_error = atoi(&buf[ARRAY_SIZE(AT_CMD_CMS_STR) - 1]);
+	} else {
+		last_error = 0;
+	}
+
+	if (tmpstr) {
+		buf[tmpstr - buf] = '\0';
+	}
+
+	return return_code;
+}
+
+static void callback_worker(struct k_work *item)
+{
+	struct callback_work_item *data =
+		CONTAINER_OF(item, struct callback_work_item, work);
+
+	if ((data != NULL) && (data->callback != NULL)) {
+		data->callback(data->data);
+	}
+
+	k_free(data);
+}
+
+
+static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
+{
+	int bytes_read;
+	char buf[RESPONSE_MAX_LEN];
+	enum at_cmd_rc return_code;
+	bool write_with_response = false;
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	LOG_DBG("AT socket thread started");
+
+	for (;;) {
+		bytes_read = recv(common_socket_fd, buf, sizeof(buf), 0);
+		write_with_response = false;
+
+		if (bytes_read < 0) {
+			LOG_ERR("AT socket recv failed with err %d",
+				bytes_read);
+
+			if (errno == -EAGAIN) {
```

Comment:
So, I should assume that EAGAIN is never returned. But should I assume that if recv return a value smaller than 0 that I am in a unrecoverable state? Or should I try to re-open socket? And in that case should I then try to close the socket first? And lastly, will I ever end up in a detectable state that is unrecoverable?

This code is from when the AT socket had the old behavior. I am not sure about the call to open_socket, I think that can be based on an misunderstanding when I was querying Glenn about socket failure mitigations.

Reply:
You can check with Glenn, but I'm pretty sure nrf socket will no longer return EAGAIN for blocking sockets which we use here, as this was incorrect behavior.

Ideally, we should try to recover (close and reopen the socket, yes you need to close it first) in case of any error, and make it fatal only if reopening fails. Personally though I think that if  we are tight on schedule, making it fatal in case of any error is good enough for now.

Reply:
Glenn was a bit unsure about it, but we agreed to treat the code as if it's fixed. When I test it it seems that that is the case, so I removed the EAGAIN test. And added code that will try to close, and the open a new socket if a read fails, ie. bytes_read < 0.